src/platforms/esp32/components/avm_builtins/uart_driver.c

```diff
@@ -368,37 +372,68 @@ static void uart_driver_do_write(Context *ctx, term msg)
     }
 
     term result_tuple = term_alloc_tuple(2, ctx);
-    term_put_tuple_element(result_tuple, 0, ref);
+    term_put_tuple_element(result_tuple, 0, term_from_ref_ticks(ref_ticks, ctx));
     term_put_tuple_element(result_tuple, 1, OK_ATOM);
 
     send_message(pid, result_tuple, glb);
 }
 
+static void uart_driver_do_close(Context *ctx, term msg)
+{
+    GlobalContext *glb = ctx->global;
+    struct UARTData *uart_data = ctx->platform_data;
+
+    term pid = term_get_tuple_element(msg, 0);
+    term ref = term_get_tuple_element(msg, 1);
+    uint64_t ref_ticks = term_to_ref_ticks(ref);
+
+    list_remove(&uart_data->listener.listeners_list_head);
+
+    term result_tuple = term_alloc_tuple(2, ctx);
+    term_put_tuple_element(result_tuple, 0, term_from_ref_ticks(ref_ticks, ctx));
+    term_put_tuple_element(result_tuple, 1, OK_ATOM);
+    send_message(pid, result_tuple, glb);
+
+    esp_err_t err = uart_driver_delete(uart_data->uart_num);
+    if (err != ESP_OK) {
+        ESP_LOGW(TAG, "Failed to delete UART driver.  err=%i\n", err);
+    }
+
+    free(uart_data);
+}
+
 static void uart_driver_consume_mailbox(Context *ctx)
 {
-    while (!list_is_empty(&ctx->mailbox)) {
```

Comment:
I'm not sure here, I mean don't we end up building up a queue that keeps growing, while we process just one message every time it wakes up?
What are you trying to accomplish here?

Reply:
If one of the messages is a close message, then the loop, as written, will crash because the context will be destroyed.  

If we were to terminate the loop and leave the rest of the messages unprocessed, then any subsequent requests would time out.

I think the better approach is to process the messages one at a time, and if any messages are delivered to the port after a close, then the process will be terminated, and (per the port implementation), the calling process will receive an `{error, noproc}`, as intended.  I think that is a lot more clear than a timeout, or worse, a process hang if a timeout was not specified.

Reply:
Put the loop back in but modified the code so that any messages received after the close will result in sending the calling client an `{error, no_proc}`